LLM Observability docs (#22668)
DataDog/documentation: content/en/tracing/llm_observability/tracing_llm_applications.md

---
title: Tracing LLM Applications
kind: guide
---
<div class="alert alert-info">LLM Observability is in public beta.</a></div>

## Overview

Your application can submit data to LLM Observability in two ways: with Datadog's [Python SDK][12], or with the [LLM Observability API][13].

> Each request fulfilled by your application is represented as a trace on the [LLM Observability traces list page][2] in Datadog:

Reviewer: ```suggestion
Each request fulfilled by your LLM application is represented as a trace on the [**Traces List** page][2] in Datadog:
```

> {{< img src="tracing/llm_observability/llm-observability-overview.png" alt="An LLM Observability trace displaying each span of a request" style="width:90%;" >}}

Reviewer: ```suggestion
{{< img src="tracing/llm_observability/llm-observability-overview.png" alt="An LLM Observability trace displaying each span of a request" style="width:100%;" >}}
```

A given trace contains spans representing each choice made by an agent or each step of a workflow. A *span* represents some unit of work that your application is performing. Spans have a start time, duration, name, tags, and attributes.

Multiple spans combine to form a trace, and a *root span* is the first span in a trace.

> A trace can contain several kinds of spans: Agent, LLM, Workflow, and so on. The *span kind* categorizes the type of work the span is performing. Different span kinds have different parent-child relationships, and only some kinds can be the root span of a trace. For details, see [Span Kinds][4].

Reviewer 2: would be great to include this as well so that new users are given a general walkthrough of what this looks like

> A trace can contain several kinds of spans: Agent, LLM, Workflow, and so on. The *span kind* categorizes the type of work the span is performing. Different span kinds have different parent-child relationships, and only some kinds can be the root span of a trace. For details, see [Span Kinds][4].

Author: @barieom, this information is included on the index page, which I would argue is where it belongs -- this guide page already has some duplication against the overview just for the sake of making sense on its own, but it's mostly intended to dig into more detailed ideas that would be too granular for the overview, like span kinds and the different specific actions you can take with the SDK and API.

Do you have suggestions for how I should flesh out this part of the overview more? I do list each type and explain what it is, but maybe you would like them under headers so it's more obvious?

> A trace can contain several kinds of spans: Agent, LLM, Workflow, and so on. The *span kind* categorizes the type of work the span is performing. Different span kinds have different parent-child relationships, and only some kinds can be the root span of a trace. For details, see [Span Kinds][4].

Reviewer 2: I would still include the short primers on[ LLM Inference Monitoring](https://docs.google.com/document/d/1o3mp-HCtNIVQUrZLOVzjrzW_GhGAoRQvieSba9oKYQY/edit#heading=h.3z1mp82tditb), [LLM Workflow Monitoring](https://docs.google.com/document/d/1o3mp-HCtNIVQUrZLOVzjrzW_GhGAoRQvieSba9oKYQY/edit#heading=h.8tpljfhuzumx), and [LLM Agent Monitoring](https://docs.google.com/document/d/1o3mp-HCtNIVQUrZLOVzjrzW_GhGAoRQvieSba9oKYQY/edit#heading=h.y15zcq73g5w0) as bullet points before `For details, see [Span Kinds][4].` sentence since we continue to get feedback from design partners around having more examples and context as these concepts are fairly new

## Instrument an LLM application

<div class="alert alert-info">This guide uses the Python SDK. If your application is not written in Python, you can complete the steps below with API requests instead of SDK function calls.</a></div>

> To trace a given LLM application:

Reviewer: ```suggestion
To trace an LLM application:
```

> 1. [Install the Python SDK][1].
> 1. [Start your application with the required env variables][5], including your Datadog API key. If you don't have an API key, you can [create one in Datadog][3].

Reviewer: ```suggestion
1. [Install the LLM Observability SDK for Python][1].
1. [Start your application with the required environment variables][5], including your Datadog API key. If you don't have an API key, you can [create one in Datadog][3].
```

> 1. In your code, [use the SDK to create spans](#create-spans) representing your application's tasks.

Reviewer: ```suggestion
1. In your code, [use the SDK to create spans](#span-creation-example) representing your application's tasks.
```

    - See the span creation example below.
    - For additional examples and detailed usage, see the [Quickstart][10] and the [SDK documentation for tracing spans][11]. 
1. [Annotate your spans][7] with input data, output data, metadata (such as `max_tokens`), and key-value tags (such as `version:1.0.0`).

> 1. Explore the resulting traces on the [traces list page][2].

Reviewer: ```suggestion
1. Explore the resulting traces on the [**Traces List** page][2].
```

Optionally, you can also:

- Associate multiple interactions with one user by specifying a `session_id`. See [Tracking user sessions][6] in the SDK documentation.
- [Persist a span between contexts or scopes][8] by manually starting and stopping it.
- [Override the name of the LLM application with a different name][9] when starting a root span, which can be useful for differentiating between services or running an experiment.

> 

Reviewer 2: ```suggestion
- If you are collecting any feedback from your end users in your LLM application (e.g. thumbs up/thumbs down, rating experience out of 1 to 5, etc), you can submit [custom evaluation metrics via the API](/llm_observability/api/?tab=model#eval-metrics-api) so that you can visualize these feedback in your traces and also in the cluster view, where you can monitor any patterns in different topics of your LLM applications against these metrics.
```

Reviewer 2: could we make sure this is included? thanks :)

### Span creation example

> To create a span, use the SDK's `LLMObs.<SPAN_KIND>()` as a context manager, replacing `<SPAN_KIND>` with the desired [span kind][4]. The example below creates a workflow span:

Reviewer: ```suggestion
To create a span, use the LLM Observability SDK for Python's `LLMObs.<SPAN_KIND>()` as a context manager, replacing `<SPAN_KIND>` with the desired [span kind][4]. 

The example below creates a workflow span:
```
